Throttle `mousemove` events to one per rendering update

whsieh opened the pull request:
#### a10c4813fe62020d18ca2073c2e687a9f86279af
<pre>
Throttle `mousemove` events to one per rendering update
<a href="https://bugs.webkit.org/show_bug.cgi?id=259408">https://bugs.webkit.org/show_bug.cgi?id=259408</a>
rdar://110921187

Reviewed by Tim Horton and Simon Fraser.

Throttle the mousemove event dispatch rate to a maximum of 1 per rendering update, if the user isn&apos;t
clicking or dragging. See below for more details.

Test:   MouseEventTests.CoalesceMouseMoveEvents
        MouseEventTests.ProcessSwapWithDeferredMouseMoveEventCompletion

* LayoutTests/fast/selectors/style-invalidation-hover-change-descendants-expected.txt:
* LayoutTests/fast/selectors/style-invalidation-hover-change-descendants.html:
* LayoutTests/fast/selectors/style-invalidation-hover-change-siblings-expected.txt:
* LayoutTests/fast/selectors/style-invalidation-hover-change-siblings.html:

Adjust a couple of layout tests that need to run checks for style invalidation state immediately
after handling the mousemove event, without waiting for any further IPC messages to be dispatched.
Ensure this by moving the test logic into one-shot `mousemove` event listeners; we also take the
opportunity to modernize the test a bit by using `js-test.js` and `testPassed` / `testFailed`.

* Source/WebKit/Shared/WebEvent.h:
* Source/WebKit/Shared/WebEventType.h: Copied from Source/WebKit/Shared/WebEvent.h.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handleMouseEvent):

See comments under `WebPage::mouseEvent`.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageFlushDeferredDidReceiveMouseEventForTesting):

Add a testing-only SPI hook to flush the `DidReceiveEvent` message corresponding to any `mousemove`
events that have already been handled in the web process. WebKitTestRunner uses this to ensure that
`window.eventSender` API to simulate mouse events behaves the same way as it currently does.

* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::suspendForProcessSwap):

Flush any deferred `mousemove` IPC responses here when suspending right before a process swap, so
that we won&apos;t end up trying to send a `DidReceiveEvent` message back to the UI process when the
`WebPageProxy`&apos;s `mouseEventQueue` has already been emptied, due to process swapping.

(WebKit::WebPage::mouseEvent):

When handling a `mousemove` event, rather than invoke the IPC completion handler immediately, we
instead defer the call to `WebPageProxy::DidReceiveEvent` until the end of the current rendering
update. This means that existing UI-side logic for coalescing `mousemove` events when the web
process is still handling mouse events will coalesce mousemove events until the end of the rendering
update, after which the web process will be done with current mousemove.

If a non-`mousemove` event enters the mouse event queue), we&apos;ll also tell the web process to
dispatch the deferred mousemove completion handler early so that we can process the incoming click
right away.

(WebKit::WebPage::flushDeferredDidReceiveMouseEvent):
(WebKit::WebPage::finalizeRenderingUpdate):

Clear out `m_deferredDidReceiveMouseEvent` if it was set, and dispatch the `DidReceiveEvent` message
back to the UI process for this deferred event.

(WebKit::WebPage::didCommitLoad):

Similarly, flush any deferred `DidReceiveEvent` response when committing a load.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm: Added.

Add an API test to exercise `mousemove` event coalescing (both this new completion deferral, as well
as the preexisting mechanism for coalescing events in the UI process). This test simulates moving a
mouse cursor over (300, 300) in the window, clicking, and then verifies that:

1.  There are at least 4 mouse events observed: a `mousemove` for the starting location, another
    `mousemove` for the final destination, and `mousedown` and `mouseup` events for the click.

2.  There are no more than 200 events in total (in other words, some `mousemove` events were
    coalesced).

Also, add an API test to verify that we don&apos;t hit a `MESSAGE_CHECK` due to sending a deferred
`mousemove` event completion message during a process swap.

* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::mouseMoveTo):

Use the above testing hook. We also wait for one extra web &lt;-&gt; UI process round trip here in order
to ensure that the UI process has received the message from the web process indicating that we&apos;ve
handled the `mousemove` event.

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/266341@main">https://commits.webkit.org/266341@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9d2db39c3946e680d613a842828603e12f1eb16

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15217 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15510 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13647 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14297 "1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15921 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11584 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12167 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19212 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12660 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15547 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12841 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10743 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12032 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3307 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16437 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12684 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3208,6 +3209,9 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
     UNUSED_PARAM(didRemoveEvent);
     LOG_WITH_STREAM(MouseHandling, stream << "UIProcess: " << (didRemoveEvent ? "replaced" : "enqueued") << " mouse event " << event.type() << " (queue size " << internals().mouseEventQueue.size() << ")");
 
+    if (isMouseDown)
```
Maybe any mouse event that's not a move?

whsieh replied:
Makes sense — changed to that.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3405,9 +3405,33 @@ void WebPage::mouseEvent(const WebMouseEvent& mouseEvent, std::optional<Vector<S
+
+    flushDeferredDidReceiveMouseEvent();
+
+    if (shouldDeferDidReceiveEvent)
```
I think a comment here saying why this logic exists would be helpful for our future selves.

whsieh replied:
Sounds good — added a comment explaining how this works.

whsieh committed:
Throttle `mousemove` events to one per rendering update
https://bugs.webkit.org/show_bug.cgi?id=259408
rdar://110921187

Reviewed by Tim Horton and Simon Fraser.

Throttle the mousemove event dispatch rate to a maximum of 1 per rendering update, if the user isn't
clicking or dragging. See below for more details.

Test:   MouseEventTests.CoalesceMouseMoveEvents
        MouseEventTests.ProcessSwapWithDeferredMouseMoveEventCompletion

* LayoutTests/fast/selectors/style-invalidation-hover-change-descendants-expected.txt:
* LayoutTests/fast/selectors/style-invalidation-hover-change-descendants.html:
* LayoutTests/fast/selectors/style-invalidation-hover-change-siblings-expected.txt:
* LayoutTests/fast/selectors/style-invalidation-hover-change-siblings.html:

Adjust a couple of layout tests that need to run checks for style invalidation state immediately
after handling the mousemove event, without waiting for any further IPC messages to be dispatched.
Ensure this by moving the test logic into one-shot `mousemove` event listeners; we also take the
opportunity to modernize the test a bit by using `js-test.js` and `testPassed` / `testFailed`.

* Source/WebKit/Shared/WebEvent.h:
* Source/WebKit/Shared/WebEventType.h: Copied from Source/WebKit/Shared/WebEvent.h.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handleMouseEvent):

See comments under `WebPage::mouseEvent`.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageFlushDeferredDidReceiveMouseEventForTesting):

Add a testing-only SPI hook to flush the `DidReceiveEvent` message corresponding to any `mousemove`
events that have already been handled in the web process. WebKitTestRunner uses this to ensure that
`window.eventSender` API to simulate mouse events behaves the same way as it currently does.

* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):
(WebKit::WebPage::suspendForProcessSwap):

Flush any deferred `mousemove` IPC responses here when suspending right before a process swap, so
that we won't end up trying to send a `DidReceiveEvent` message back to the UI process when the
`WebPageProxy`'s `mouseEventQueue` has already been emptied, due to process swapping.

(WebKit::WebPage::mouseEvent):

When handling a `mousemove` event, rather than invoke the IPC completion handler immediately, we
instead defer the call to `WebPageProxy::DidReceiveEvent` until the end of the current rendering
update. This means that existing UI-side logic for coalescing `mousemove` events when the web
process is still handling mouse events will coalesce mousemove events until the end of the rendering
update, after which the web process will be done with current mousemove.

If a non-`mousemove` event enters the mouse event queue), we'll also tell the web process to
dispatch the deferred mousemove completion handler early so that we can process the incoming click
right away.

(WebKit::WebPage::flushDeferredDidReceiveMouseEvent):
(WebKit::WebPage::finalizeRenderingUpdate):

Clear out `m_deferredDidReceiveMouseEvent` if it was set, and dispatch the `DidReceiveEvent` message
back to the UI process for this deferred event.

(WebKit::WebPage::didCommitLoad):

Similarly, flush any deferred `DidReceiveEvent` response when committing a load.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm: Added.

Add an API test to exercise `mousemove` event coalescing (both this new completion deferral, as well
as the preexisting mechanism for coalescing events in the UI process). This test simulates moving a
mouse cursor over (300, 300) in the window, clicking, and then verifies that:

1.  There are at least 4 mouse events observed: a `mousemove` for the starting location, another
    `mousemove` for the final destination, and `mousedown` and `mouseup` events for the click.

2.  There are no more than 200 events in total (in other words, some `mousemove` events were
    coalesced).

Also, add an API test to verify that we don't hit a `MESSAGE_CHECK` due to sending a deferred
`mousemove` event completion message during a process swap.

* Tools/WebKitTestRunner/InjectedBundle/EventSendingController.cpp:
(WTR::EventSendingController::mouseMoveTo):

Use the above testing hook. We also wait for one extra web <-> UI process round trip here in order
to ensure that the UI process has received the message from the web process indicating that we've
handled the `mousemove` event.

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/266341@main
